data/engine.py

```diff
@@ -0,0 +1,118 @@
+    return heapq.nlargest(n, courses, key=lambda course:
+                         (course.interest.rating, course.interest.count))
+
+
+class RecommendationEngine:
+    def __init__(self, sc):
+        """
+        Initialize the engine given a spark context
+        """
+        log.info('Starting up the Recommendation Engine: ')
+        self.sc = sc
+
+        # ALS algorithm needs 32 bit integer for user_id and course_id
+        # So I mapped user_id and course_id in the db to an unique integer
+        # Probably not the best solution, but this is just a prototype
+        log.info('Loading user data')
+        global user_lookup
+        user_count = 0
+        for user in m.User.objects:
+            user_lookup[str(user.id)] = user_count
+            user_count += 1
+
+        log.info('Loading course data')
+        global course_lookup
+        course_count = 0
+        for course in m.Course.objects:
+            course_lookup[course.id] = course_count
+            course_count += 1
+
+        # send list of (user_id, course_id, rating) triples to the ML algorithm
+        log.info('Loading ratings data')
+        ratings_RDD_raw = self.sc.parallelize(m.UserCourse.objects)
+        self.ratings_RDD = (ratings_RDD_raw
+                       .filter(lambda ratings:
+                               ratings.course_review.interest is not None)
+                       .map(lambda ratings:
+                            (user_lookup[str(ratings.user_id)],
+                             course_lookup[ratings.course_id],
+                             float(ratings.course_review.interest)))
+                       )
+
+        self.rank = 4
+        self.num_iterations = 10
+        self.__train(self.ratings_RDD, self.rank, self.num_iterations)
+
+    def __train(self, ratings, rank, num_iterations):
+        log.info('Training model')
+        self.model = ALS.trainImplicit(ratings,
+                                       rank,
+                                       num_iterations,
+                                       lambda_=0.05)
+        log.info('Model trained!')
+
+    def recommend_user(self, user_id, n):
+        """Recommend top n courses to the user provided"""
+        if user_id:
+            user = user_lookup.get(user_id, None)
+            if not user:
+                raise Exception('User %s is not in the database' % user_id)
+        else:
+            raise Exception('Please provider a user id')
+
+        top_courses = []
+        user_with_ratings = zip(*(self.ratings_RDD.collect()))[0]
+        if user in user_with_ratings:
+            count = len([i for i in user_with_ratings if i == user])
+            top_ratings = self.model.recommendProducts(user, n + count)
+            inverse_course_lookup = {v: k for k, v in course_lookup.items()}
+            user_courses = [course_lookup[uc.course_id] for uc in
+                            m.User.objects(id=user_id).
```

Comment:
Indent here to show it's part of the same statement? Similar to what you did for the `top_courses` assignment